tendermint/src/proposal/sign_proposal.rs

```diff
@@ -0,0 +1,96 @@
+use bytes::BufMut;
+use std::convert::TryFrom;
+use tendermint_proto::privval::RemoteSignerError;
+use tendermint_proto::privval::SignProposalRequest as RawSignProposalRequest;
+use tendermint_proto::privval::SignedProposalResponse as RawSignedProposalResponse;
+use tendermint_proto::DomainType;
+use tendermint_proto::Error as DomainTypeError;
+
+/// SignProposalRequest is a request to sign a proposal
+#[derive(Clone, PartialEq, Debug)]
+pub struct SignProposalRequest {
+    /// Proposal
+    pub proposal: Proposal,
+    /// Chain ID
+    pub chain_id: ChainId,
+}
+
+impl DomainType<RawSignProposalRequest> for SignProposalRequest {}
+impl DomainType<RawSignedProposalResponse> for SignedProposalResponse {}
+
+impl TryFrom<RawSignProposalRequest> for SignProposalRequest {
+    type Error = Error;
+
+    fn try_from(value: RawSignProposalRequest) -> Result<Self, Self::Error> {
+        if value.proposal.is_none() {
+            return Err(Kind::NoProposalFound.into());
+        }
+        Ok(SignProposalRequest {
+            proposal: Proposal::try_from(value.proposal.unwrap())?,
+            chain_id: ChainId::try_from(value.chain_id).unwrap(),
+        })
+    }
+}
+
+impl From<SignProposalRequest> for RawSignProposalRequest {
+    fn from(value: SignProposalRequest) -> Self {
+        RawSignProposalRequest {
+            proposal: Some(value.proposal.into()),
+            chain_id: value.chain_id.to_string(),
+        }
+    }
+}
+
+impl SignProposalRequest {
+    /// Create signable bytes from Proposal.
+    pub fn to_signable_bytes<B>(&self, sign_bytes: &mut B) -> Result<bool, DomainTypeError>
+    where
+        B: BufMut,
+    {
+        self.proposal
+            .to_signable_bytes(self.chain_id.clone(), sign_bytes)
+    }
+
+    /// Create signable vector from Proposal.
+    pub fn to_signable_vec(&self) -> Result<Vec<u8>, DomainTypeError> {
+        self.proposal.to_signable_vec(self.chain_id.clone())
+    }
+}
+
+/// SignedProposalResponse is response containing a signed proposal or an error
+#[derive(Clone, PartialEq)]
+pub struct SignedProposalResponse {
+    /// Proposal
+    pub proposal: Option<Proposal>,
+    /// Response error
+    pub error: Option<RemoteSignerError>,
+}
+
+impl TryFrom<RawSignedProposalResponse> for SignedProposalResponse {
+    type Error = Error;
+
+    fn try_from(value: RawSignedProposalResponse) -> Result<Self, Self::Error> {
+        Ok(SignedProposalResponse {
+            proposal: match value.proposal {
+                None => None,
+                Some(proposal) => Some(Proposal::try_from(proposal)?),
+            },
+            error: value.error,
+        })
+    }
+}
+
+impl From<SignedProposalResponse> for RawSignedProposalResponse {
+    fn from(value: SignedProposalResponse) -> Self {
+        RawSignedProposalResponse {
+            proposal: match value.proposal {
+                None => None,
+                Some(proposal) => Some(proposal.into()),
+            },
```

Comment:
This can likely be simplified to:

```suggestion
            proposal: value.proposal.map(Into::into),
```

or perhaps like this if rustc fails to infer the appropriate `Into` instance:

```suggestion
            proposal: value.proposal.map(TargetType::from),
```

(after substituting the actual target type for `TargetType`)

Reply:
Yup, another example of late-night coding.
I exchanged it to the currently used `.map(|p| p.into())` for the next commit. I'll do another sweep where I want to change all of these into `.map(into::Into)` which is what I wanted to do in the first place, I just didn't know I have to explicitly call the trait to be able to do it...

Reply:
Covered in the third installment of fixes.